das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+//
+// Copyright 2022, Offchain Labs, Inc. All rights reserved.
+//
+
+package das
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/bits"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/blsSignatures"
+)
+
+type AggregatorConfig struct {
+	assumedHonest   int
+	retentionPeriod time.Duration
+}
+
+type Aggregator struct {
+	config   AggregatorConfig
+	services []serviceDetails
+
+	/// calculated fields
+	requiredServicesForStore       int
+	maxAllowedServiceStoreFailures int
+}
+
+type serviceDetails struct {
+	service    DataAvailabilityService
+	pubKey     blsSignatures.PublicKey
+	signerMask uint64
+}
+
+func NewAggregator(config AggregatorConfig, services []serviceDetails) *Aggregator {
+	return &Aggregator{
+		config:                         config,
+		services:                       services,
+		requiredServicesForStore:       len(services) + 1 - config.assumedHonest,
+		maxAllowedServiceStoreFailures: config.assumedHonest - 1,
+	}
+}
+
+// Retrieve calls  on each backend DAS in parallel and returns immediately on the
+// first successful response where the data matches the requested hash. Otherwise
+// if all requests fail or if its context is canceled (eg via DeadlineWrapper) then
+// it returns an error.
+func (a *Aggregator) Retrieve(ctx context.Context, cert []byte) ([]byte, error) {
+	requestedCert, _, err := arbstate.DeserializeDASCertFrom(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cert is the aggregate cert, validate it against DAS public keys
+	var servicesThatSignedCert []serviceDetails
+	var pubKeys []blsSignatures.PublicKey
+	for _, d := range a.services {
+		if requestedCert.SignersMask&d.signerMask != 0 {
+			servicesThatSignedCert = append(servicesThatSignedCert, d)
+			pubKeys = append(pubKeys, d.pubKey)
+		}
+	}
+	signedBlob := serializeSignableFields(*requestedCert)
```

Comment:
Should verify here that `len(servicesThatSignedCert)` is large enough.

Reply:
Added the check.